stl/src/syserror_import_lib.cpp

```diff
@@ -40,15 +40,35 @@ extern "C" {
     // convert to name of Windows error, return 0 for failure, otherwise return number of chars in buffer
     // __std_system_error_deallocate_message should be called even if 0 is returned
     // pre: *_Ptr_str == nullptr
-    DWORD _Lang_id;
-    const int _Ret = GetLocaleInfoEx(LOCALE_NAME_SYSTEM_DEFAULT, LOCALE_ILANGUAGE | LOCALE_RETURN_NUMBER,
-        reinterpret_cast<LPWSTR>(&_Lang_id), sizeof(_Lang_id) / sizeof(wchar_t));
-    if (_Ret == 0) {
-        _Lang_id = 0;
+
+    // We start by requesting US English for system_category() messages. (See GH-2451 and GH-3254 for the history.)
+    // This is consistent with generic_category(), which uses a table of US English strings in the STL.
+    // In general, system_error messages aren't directly useful to end-users - they're meant for programmer-users.
+    // Of course, the programmer-user might not speak US English, but machine translation of the message
+    // (and the numeric value of the error code) should help them understand the error.
+
+    constexpr auto _Flags = FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS;
+
+    DWORD _Lang_id = 0x0409; // 1033 decimal, "en-US" locale
+
+    auto _Chars = FormatMessageA(_Flags, nullptr, _Message_id, _Lang_id, reinterpret_cast<char*>(_Ptr_str), 0, nullptr);
+
+    if (_Chars == 0) {
+        // If FormatMessageA() failed for any reason, cleanup and restore this function's precondition.
+        LocalFree(*_Ptr_str);
+        *_Ptr_str = nullptr;
+
+        // Attempt to get the system locale's language ID.
+        const int _Ret = GetLocaleInfoEx(LOCALE_NAME_SYSTEM_DEFAULT, LOCALE_ILANGUAGE | LOCALE_RETURN_NUMBER,
+            reinterpret_cast<LPWSTR>(&_Lang_id), sizeof(_Lang_id) / sizeof(wchar_t));
+
+        // If we can't get the system locale, the final fallback is FormatMessageA()'s behavior for language ID 0.
+        if (_Ret == 0) {
+            _Lang_id = 0;

```

Comment:
I don't believe this handles the case where we can get the system locale, but calling FormatMessageA would fail.

Reply:
For example, what happens in a system with a `fr-CA` system locale, but `fr-FR` Windows display language? Is this a scenario that one could trigger on a single language install?

Reply:
I could add logic, but I'd really want it to be motivated by an actual test run.

Reply:
I'll change to a three-step process, thanks.